std/utf.d

```diff
@@ -1164,8 +1164,10 @@ package template codeUnitLimit(S)
  * Returns:
  *      decoded character
  */
-private dchar decodeImpl(bool canIndex, UseReplacementDchar useReplacementDchar = UseReplacementDchar.no, S)(auto ref S str, ref size_t index)
-    if (is(S : const char[]) || (isInputRange!S && is(Unqual!(ElementEncodingType!S) == char)))
+private RV decodeImpl(bool canIndex, UseReplacementDchar useReplacementDchar, RV = dchar, S)
```

Comment:
It's `RV` for `ReturnValue` and is much shorter in an already very long function signature. I could certainly change it to `ReturnType`, but this is private, so it's not like anyone is going to see it in the docs, just if they're maintaining `decode` and friends that use `decodeImpl`. `CharType` wouldn't really make sense, since it's either `dchar` or `bool`, and `bool` isn't a character type.

In same ways, what I think the right thing to do is to have a 3 value enum instead of `UseReplacementDchar` where it's equivalent to yes, no, and onlyValidate, and then the return type would be `auto` an dthere would be no `RV` template parameter. But then some of the code calling `decodeImpl` would be forced to translate `UseReplacementDchar` into this other, private enum type, which gets kind of ugly. But what's here is kind of ugly already, so I don't know. Reusing this functionality without duplicating it has definitely turned out to be rather "interesting" as we've evolved what we've tried to do with it.